### ISSUE TITLE
Apple auth email is null (data apple signin and firebase email null)

### ISSUE THREAD

**Issue by marquina-abreu** [invertase/react-native-apple-authentication#187]
- react-native: 0.63.3
- @intervase/react-native-apple-authentication: 2.1.0
- react-native-firebase: 5.6.0

### TESTED
- Tested on device real iPhone 12 pro max (iOS 14.3)
- Xcode 12.3

### EMAIL is NULL

```javascript
export const onAppleButtonPress = async () => {
  // performs login request
  try {
    const appleAuthRequestResponse = await appleAuth.performRequest({
      requestedOperation: appleAuth.Operation.LOGIN,
      requestedScopes: [appleAuth.Scope.EMAIL, appleAuth.Scope.FULL_NAME],
    });

    // get current authentication state for user
    const credentialState = await appleAuth.getCredentialStateForUser(
      appleAuthRequestResponse.user,
    );
    const dataDecode = jwtDecode(appleAuthRequestResponse.identityToken);
    console.log('DATA DECODEE ',dataDecode);
    console.log('USER LOG APPLEEEE ', appleAuthRequestResponse);

    // use credentialState response to ensure the user is authenticated
    const { identityToken, nonce } = appleAuthRequestResponse;
    if (identityToken) {
      // e.g. sign in with Firebase Auth using `nonce` & `identityToken`
      const appleCredential = firebaseFather.auth.AppleAuthProvider.credential(
        identityToken,
        nonce,
      );
      const userCredential = await authFirebase.signInWithCredential(appleCredential);
      Flux.dispatchEvent(LOGIN_APPLE);
      // console.log('userCredential with Applee ', userCredential);
      console.log('Firebase authenticated via Apple, UID: ', userCredential.user.uid);
      if (credentialState === appleAuth.State.AUTHORIZED) {
        console.log(' insidee of AUTHORIZED success ');
        // user is authenticated
      }
    } else {
      const error = {
        message: 'There was an error starting with sign in apple.',
      };
      Flux.dispatchEvent(LOGIN_APPLE_ERROR, error);
    }
  } catch (error) {
    console.log('logs of Error ', error);
    if (error.code === appleAuth.Error.CANCELED) {
      error.message = 'Sign in Apple was canceled';
      Flux.dispatchEvent(LOGIN_APPLE_ERROR, error);
    } else {
      error.message = 'There was an error starting with sign in apple';
      Flux.dispatchEvent(LOGIN_APPLE_ERROR, error);
    }
  }
};
```

### LOGS

<img width="345" alt="Screen Shot 2021-01-19 at 10 27 40 AM" src="https://user-images.githubusercontent.com/34238806/105048303-c803d300-5a41-11eb-9500-8d9057bcddb7.png">

### Firebase

The user is saved without email, so I cannot continue the registration or linking of users

<img width="825" alt="Screen Shot 2021-01-19 at 10 28 03 AM" src="https://user-images.githubusercontent.com/34238806/105048335-d225d180-5a41-11eb-9afa-bdc0658c9b1b.png">

**Comment by marquina-abreu** [comment]
@mikehardy hi, can you help me?

**Comment by mikehardy** [comment]
If you can reproduce with the example included here, sure

**Comment by alacret** [comment]
```
eyJraWQiOiI4NkQ4OEtmIiwiYWxnIjoiUlMyNTYifQ.eyJpc3MiOiJodHRwczovL2FwcGxlaWQuYXBwbGUuY29tIiwiYXVkIjoiY29tLm9rcm9vIiwiZXhwIjoxNjExMTU3NjcwLCJpYXQiOjE2MTEwNzEyNzAsInN1YiI6IjAwMTU4OC5kNThkODQzMDFlNzg0NTkxOTEwNDBhY2I5NDdkZWZlZi4wMzA5Iiwibm9uY2UiOiJjZmYxOTAyOGIxMjRlMzY3MTNjZGI0NDYzMWNhYTRkMWVjODAxNjRmYTk2YTJlN2E2ZjljMjhjY2U1NjY0YWIwIiwiY19oYXNoIjoicFVuY2FESWk5WkRERjNURDJxSUh4USIsImF1dGhfdGltZSI6MTYxMTA3MTI3MCwibm9uY2Vfc3VwcG9ydGVkIjp0cnVlfQ.N1PJjCy8KnqzQWTiWEVZXZHrHZuvBDIX5FvgZEljfo5fE8UPoZtx0HfUBmJoUzgeWWANT4iY9iZIGoD9QnBbUsi1qgm_L8dmRNEWNb8P17lOAiDeOMUEfYMoVgwOsFhXfA4Tm-ZgssxmcoOonUpqTwvubT26gpkFArzeAr4QxitB3Het2AnFyTzLNx9sKQbzbP8HHyBeuZpvPGeIkUmOODdShbj6duS_2qvEJLxhJHNOeYoDW9Tjj_Pcx2i0FQJl_ZM7qnJn8qH4nvnMQ65VyNWDqaUxLlIB4HtGSks_hF9xEZ7et4cu7brGbOYvpyZudpP3qNP-hM0biaIopCU4xA
```

<img width="1680" alt="Screen Shot 2021-01-19 at 10 50 18 AM" src="https://user-images.githubusercontent.com/918895/105058355-216d0180-5a44-11eb-8427-5fb7ea63cdd9.png">
![Screen Shot 2021-01-19 at 10 50 18 AM (2)](https://user-images.githubusercontent.com/918895/105058362-229e2e80-5a44-11eb-88bc-1e5ec59de62f.png)

**Comment by marquina-abreu** [comment]
yes @mikehardy  @alacret 

the onAppleButtonPress function is what I call when I login in the real device, but it never brings me the email and I also tried to use a `jwtDecode (appleAuthRequestResponse.identityToken);` and here the email doesn't arrive either 

but this information arrives to firebase auth but firebase doesn't return the email either and in fact the user is inserted but without email in firebase

**Comment by mikehardy** [comment]
How did it go when you reproduce with the example?

**Comment by marquina-abreu** [comment]
well, but 'appleAuthRequestResponse' never returns my email and neither does identityToken contain email @mikehardy

**Comment by marquina-abreu** [comment]
so firebase doesn't recognize the email from the parameters that passed it to the function `firebase.auth.AppleAuthProvider.credential(identityToken, nonce);` and so I can't continue the process

**Comment by mikehardy** [comment]
So you are telling me, if I integrate this module with react-native-firebase right now and do an apple signing (for instance on an ios13 simulator) that a new user won't work because email is not present? If you answer that this is correct I will verify this next time I'm near a Mac.

**Comment by marquina-abreu** [comment]
exactly but in my case I tried it from a real device, the iPhone 12 pro max, not from a simulator because you clearly indicate that in a simulator gives error @mikehardy

**Comment by mikehardy** [comment]
I clearly indicated ios13 works great. Ios14 fails. Okay I will test, maybe Apple changed something

**Comment by alacret** [comment]
> I clearly indicated ios13 works great. Ios14 fails. Okay I will test, maybe Apple changed something

@mifi Thanks let us know if we can do anything to help

**Comment by marquina-abreu** [comment]
@mikehardy  ahm ok ok yes, my iPhone is iOS 14.3 with that was the one I tested, thanks let us know if we can do anything to help

**Comment by nericode** [comment]
Hello, same issue in iOS 14.3, `email is null`

**Comment by mikehardy** [comment]
The example app works perfectly for me on ios13 emulator, watch the console output.
I get my email in there, and the identityToken contains it as well.
I'll check on my phone later but this has a lower priority for me now.

People reporting problems: check out the repository, `yarn && cd example && yarn && cd ios && pod install`, then from example dir `yarn start` in one terminal and create an emulator profile named `iPhone 11 ios13` that is (you guessed it) an iPhone 11 but configured for iOS 13.7 (as of this typing?), then run `yarn run:ios13`

It should work.

We have had lots of users with issues getting emails before. Not one has been a problem with the module yet, so the burden of proof is on reporters here to reproduce this with the example watching the console in the simulator and with a real device plugged in / example launched from xcode

**Comment by ChanghyeonYoon** [comment]
Apple is give the data only one time. try another account.

**Comment by marquina-abreu** [comment]
I did what @ChanghyeonYoon  indicated and the login worked, I do not understand what happened, I remember that the last time I logged in was on an iPhone 8 months ago in the same app with the same apple id and then I logged in recently with the new iPhone and that's when I got the error. thank you very much! @mikehardy  @ChanghyeonYoon

**Comment by amitbravo** [comment]
@ marquina-abreu if you once try to register with apple id and you allow to share your email id and name, things go fine if you do not cancel from the pop up, once you cancel the process name and email goes null all the time until somehow as @ ChanghyeonYoon stated you do not reset it so if there any method to revoke/reset if we cancel then it can work fine.

**Comment by dprajapati1179** [comment]
How to get an email and name each time if allowed in the first attempt ?

**Comment by Romick2005** [comment]
Just save it on your side (DB?) with received apple id. So for the next login you will get only apple id, but you can pick email and name from stored on your side first login. So no, it is not possible to email and name on every 2, 3 ... n login attemp.

**Comment by Romick2005** [comment]
Just save it on your side (DB?) with received apple id. So for the next login you will get only apple id, but you can pick email and name from stored on your side first login. So no, it is not possible to email and name on every 2, 3 ... n login attempt.

**Comment by juniordeleon013** [comment]
Is this Apple ID unique for every user?

**Comment by Romick2005** [comment]
Yes. That is how you should map apple id with the user!

**Comment by jaesam** [comment]
@Romick2005 when you said apple id, did you mean apple id token? or the actual apple id that the user used to login?

**Comment by Romick2005** [comment]
> @Romick2005 when you said apple id, did you mean apple id token? or the actual apple id that the user used to login? 

It is apple id that you will get in response on every login attempt and it is number type. You will not get actual Apple id used to login, because that usually is an email, but user can hide it using apple relay email. And it's type is string.

**Comment by ElixirMike** [comment]
I'm running into this issue, but I can't quite tell what the solution is? I'm developing the logic and I had attempted to login with apple, and only after looking further did I see the email was null. I read that you can use the device settings to find the AppleID and find your app and choose to stop using AppleID, which should reset it...so that the next time you attempt to login, you theoretically would get the email, but that's not happening for me...it's always null.

On first attempt using appleID, should it provide the name and email?
If I missed that window is there anwyay to reset it, so I can try again or do I need to create a whole new apple id?

**Comment by ZeeshanAhmadKhalil** [comment]
> 

tried this still getting null email and username

**Comment by felixkrautschuk** [comment]
I am not developing react native apps and therefore I am not using this plugin, but I also came across this issue while developing our NativeScript app for iOS.

For the first manual signin, the identityToken contains the email as expected. But when the user then goes to iOS settings and then "Stop using Apple ID" for our app, any future signin with that apple id will return null for the email and the identityToken is a lot shorter and does not contain the email info anymore.

How am I able to access the email for that apple id again??

**Comment by felixkrautschuk** [comment]
Never mind, my problem is only occurring when testing with Simulator... when running the app on real device, apple sign in is working as expected, even after the user stops using the apple id in the iOS security settings.

**Comment by RafikMk** [comment]
Hey, just so you know, Apple only shares user data the first time they log in.
Wanna try again? You might want to use a different Apple account.

Or, if you prefer to use the same account, here's a quick fix:

- Head over to the Settings on your iPhone.
- Tap on your Apple ID, then go to 'Password & Security'.
- Find 'Apps using Apple ID' and select your app.
- Hit 'Stop Using Apple ID' to disconnect it.

This resets your login details.
When you log in next, it'll be like a first-time thing and you should get the data again.

**Comment by e1016** [comment]
> Hey, just so you know, Apple only shares user data the first time they log in. Wanna try again? You might want to use a different Apple account.
> 
> Or, if you prefer to use the same account, here's a quick fix:
> 
> * Head over to the Settings on your iPhone.
> * Tap on your Apple ID, then go to 'Password & Security'.
> * Find 'Apps using Apple ID' and select your app.
> * Hit 'Stop Using Apple ID' to disconnect it.
> 
> This resets your login details. When you log in next, it'll be like a first-time thing and you should get the data again.

This is not working for me in simulator (and I dont have en iphone) actually if you sign again after "stop using apple id" yu will see that your app has received nothing from apple
<img width="454" alt="Captura de pantalla 2024-04-10 a la(s) 7 47 23 a m" src="https://github.com/invertase/react-native-apple-authentication/assets/29954550/e82301b1-c4a4-4f55-bffc-23f2029b7924">
